// transmuter/index.js
var lodash = {
        utils : {
            parseInt: require('lodash-node/modern/utilities/parseInt')
        }
    },
    debug = function (message) { console.log(message); },
    Transmuter;

/**
 *  This class should be used to transform data from one type to another
 *  Like turning "0.00" (string) into a decimal/float 0.00 or any other type of
 *  changes like that
 */

Transmuter = {

    fromJSONString: function fromJSONString(data) {
        return JSON.parse(data);
    },

    toJSONString: function toJSONString(data) {
        return JSON.stringify(data);
    },

    toBoolean: function toBoolean(value) {
        // type = type || 'numeric';
        //
        // if (type === 'numeric') {
        //     if(value) {
        //         return 1;
        //     } else {
        //         return 0;
        //     }
        // } else {
            return Boolean(value);
        // }
    },

    toString: function toString(string) {
        return string.toString() || '';
    },

    toFloat: function toFloat(number) {
        return parseFloat(number) || 0.00;
    },

    toInt: function toInt(number) {
        return lodash.utils.parseInt(number) || 0;
    }
};

module.exports = Transmuter;
